## Matrix Factorization and Stochastic Gradient Descent


### step one: write a dot product function

15xp

In [35]:
def dot(a, b):
    total = 0
    # To be done
    return total

a = [2, 4, 6]
b = [3, 6, 9]
dot(a, b)


84

should return 84

### now some additional functions  we will need:

In [46]:
p = [[ 1.3756739,   1.82247159],
 [ 1.16901533,  1.39862014],
 [ 2.08942885, -0.65458002],
 [ 1.69334605, -0.48281951],
 [ 1.89681673, -0.14079464]]

q = [[1.11071814, 0.72430544,2.60828792, 2.05752625,],
    [1.90348475, 1.06266696, 0.29271036, -1.00670788]]


q2 = [[ 1.11071814,  1.90348475],
 [ 0.72430544,  1.06266696],
 [ 2.60828792,  0.29271036],
 [ 2.05752625, -1.00670788]]

def transpose(a):
    result = [[a[y][x] for y in range(len(a))] for x in range(len(a[0]))]
    
    return result

def matMult(a, b):
    #matrix multiplication
    b  = transpose(b)
    result = [[0 for y in range(len(b))] for x in range(len(a))]
    for x in range(len(a)):
        for y in range(len(b)):
            result[x][y] = dot(a[x], b[y])
    return result

def prettyPrintMatrix(a):
  for x in a:
    print (x)

#    my test code
r = matMult(p, q)
prettyPrintMatrix(r)
print('\n\n\n')
prettyPrintMatrix(transpose(r))




[4.997032834327799, 2.9330884336676823, 4.121609930427961, 0.995788649960746]
[3.960698640501951, 2.3329915753309693, 3.4585191682164638, 0.997277812062709]
[1.0747834402096434, 0.8177841226178048, 5.258229675851485, 4.95802557060687]
[0.9617906007448743, 0.7134234148971222, 4.275407774007592, 3.9701621535435514]
[1.8388283001447419, 1.2242568641489167, 4.906232113552432, 4.044489286963926]




[4.997032834327799, 3.960698640501951, 1.0747834402096434, 0.9617906007448743, 1.8388283001447419]
[2.9330884336676823, 2.3329915753309693, 0.8177841226178048, 0.7134234148971222, 1.2242568641489167]
[4.121609930427961, 3.4585191682164638, 5.258229675851485, 4.275407774007592, 4.906232113552432]
[0.995788649960746, 0.997277812062709, 4.95802557060687, 3.9701621535435514, 4.044489286963926]


####my results
     [4.997032834327799, 2.9330884336676823, 4.121609930427961, 0.995788649960746]
     [3.960698640501951, 2.3329915753309693, 3.4585191682164638, 0.997277812062709]
     [1.0747834402096434, 0.8177841226178048, 5.258229675851485, 4.95802557060687]
     [0.9617906007448743, 0.7134234148971222, 4.275407774007592, 3.9701621535435514]
     [1.8388283001447419, 1.2242568641489167, 4.906232113552432, 4.044489286963926]


     [4.997032834327799, 3.960698640501951, 1.0747834402096434, 0.9617906007448743, 1.8388283001447419]
     [2.9330884336676823, 2.3329915753309693, 0.8177841226178048, 0.7134234148971222, 1.2242568641489167]
     [4.121609930427961, 3.4585191682164638, 5.258229675851485, 4.275407774007592, 4.906232113552432]
     [0.995788649960746, 0.997277812062709, 4.95802557060687, 3.9701621535435514, 4.044489286963926]


### Gradient Descent

#### finish the algorithm for 30xp

Here is the data we will use.

|Customer | Taylor Swift | Miranda Lambert | Carrie Underwood | Nicki Minaj | Ariana Grande |
|:-----------|:------:|:------:|:---------:|:------:|:--------:|
|Jake|5|-|5|2|2|
|Clara|2|-|-|4|5|
|Kelsey|5|5|5|2|-|
|Angelica|2|3|-|5|5|
|Jordyn|2|1|-|5|-|


Now you need to finish the following algorithm:

In [48]:

def matrix_factorization(R, P, Q, K, steps=5000, alpha=0.0002, beta=0.02):
    Q = transpose(Q)
    for step in range(steps):
        for i in range(len(R)):
            for j in range(len(R[i])):
                if R[i][j] > 0:
                    eij = R[i][j] - dot(P[i],transpose(Q)[j])
                    #
                    # Now you need to update the P and Q matrices
                    
        
        # don't need to alter or add anything after this line
        eR = matMult(P,Q)
        e = 0
        for i in range(len(R)):
            for j in range(len(R[i])):
                if R[i][j] > 0:
                    e = e + pow(R[i][j] - dot(P[i],transpose(Q)[j]), 2)
                    for k in range(K):
                        e = e + (beta/2) * (pow(P[i][k],2) + pow(Q[k][j],2))
        if e < 0.001:
            break
    return P, transpose(Q)



R = [
     [5, 0, 5, 2, 2],
     [2, 0, 0, 4, 5],
     [5, 5, 5, 2, 0],
     [2, 3, 0, 5, 5],
     [2, 1, 0, 5, 0]
    ]


K = 2

## these are some random arrays I created

P = [[ 0.64132372,  0.3808661 ],
     [ 0.092069,    0.82043744],
     [ 0.77184994,  0.08276139],
     [ 0.36048553,  0.05087073],
     [ 0.37160684,  0.3147877 ]]

Q = [[ 0.69314147,  0.75344936],
     [ 0.75093344,  0.36297899],
     [ 0.22309714,  0.58856832],
     [ 0.11611411,  0.30807273],
     [ 0.04559444,  0.29029353]]


nP, nQ = matrix_factorization(R, P, Q, K)
print("Our final estimate of the P matrix")
prettyPrintMatrix (nP)
print("\nOur final estimate of the Q matrix")
prettyPrintMatrix (nQ)
nR = matMult(nP, transpose(nQ))
print("\n\nOur predictions")
prettyPrintMatrix (nR)

Our final estimate of the P matrix
[2.3610176655454493, 0.4439041251268022]
[0.6137938317930417, 1.8591178978338283]
[2.326666932204386, 0.46068244286925697]
[0.8572268226166766, 2.013617527721459]
[0.32274717667725983, 2.239859945627974]

Our final estimate of the Q matrix
[2.0187048261349063, 0.40145041435638706]
[2.1223208557273714, 0.33442067280710364]
[1.8321637142068383, 1.5391839075388205]
[0.44154179777070157, 2.156216743837729]
[0.39155773901155466, 2.4160709447019313]


Our predictions
[4.944403250993033, 5.159287748514613, 5.009020981298982, 1.999641491870372, 1.9969785978395125]
[1.9854122209152159, 1.9243949085525076, 3.9860951371784163, 4.279676772028168, 4.732096460727768]
[4.881794922426067, 5.091995487043348, 4.97190973077985, 2.0206518969659273, 2.0240659083576187]
[2.53885551476764, 2.4927056921250674, 4.669907573850071, 4.720297301313576, 5.200696598954197]
[1.5507239864524245, 1.4340285341998649, 4.038882049422134, 4.972129887229999, 5.538034689605507]


### Here were my predictions. 
How did you do?

|Customer | Taylor Swift | Miranda Lambert | Carrie Underwood | Nicki Minaj | Ariana Grande |
|:-----------|:------:|:------:|:---------:|:------:|:--------:|
|Jake|4.94 | 5.159|5.009 | 1.999 | 1.997 |
|Clara|1.9854| 1.9243| 3.9860| 4.2796|4.7320|
|Kelsey|4.8817|5.0919| 4.9719| 2.0206|2.0240|
|Angelica|2.5388|2.4927| 4.6699| 4.72029| 5.2006|
|Jordyn|1.5507| 1.4340|4.0388| 4.9721|5.5380|